common/src/main/java/org/ergoplatform/appkit/AppkitProvingInterpreter.scala

```diff
@@ -53,27 +54,54 @@ class AppkitProvingInterpreter(
     .filter { case _: DLogProverInput => true case _ => false}
     .map(_.asInstanceOf[DLogProverInput].publicImage)
 
+  def addCost(currentCost: Long, delta: Long, limit: Long, msg: String): Long = {
+    val newCost = Math.addExact(currentCost, delta)
+    if (newCost > limit)
+      throw new Exception(s"Cost of transaction $newCost exceeds limit $limit: $msg")
+    newCost
+  }
+
   /**
    * @note requires `unsignedTx` and `boxesToSpend` have the same boxIds in the same order.
+   * @param baseCost the cost accumulated before this transaction
+   * @return a new signed transaction with all inputs signed and the cost of this transaction
+   *         The returned cost doesn't include `baseCost`.
    */
   def sign(unsignedTx: UnsignedErgoLikeTransaction,
            boxesToSpend: IndexedSeq[ExtendedInputBox],
            dataBoxes: IndexedSeq[ErgoBox],
-           stateContext: ErgoLikeStateContext): Try[ErgoLikeTransaction] = Try {
+           stateContext: ErgoLikeStateContext,
+           baseCost: Int): Try[(ErgoLikeTransaction, Int)] = Try {
     if (unsignedTx.inputs.length != boxesToSpend.length) throw new Exception("Not enough boxes to spend")
     if (unsignedTx.dataInputs.length != dataBoxes.length) throw new Exception("Not enough data boxes")
 
     // Cost of transaction initialization: we should read and parse all inputs and data inputs,
     // and also iterate through all outputs to check rules
-    val inputsCost = boxesToSpend.size * params.inputCost
-    val dataInputsCost = dataBoxes.size * params.dataInputCost
-    val outputsCost = unsignedTx.outputCandidates.size * params.outputCost
-    val initialCost: Long = inputsCost + dataInputsCost + outputsCost
+    val initialCost = ArithUtils.addExact(
+      CostTable.interpreterInitCost,
+      Math.multiplyExact(boxesToSpend.size, params.inputCost),
+      Math.multiplyExact(dataBoxes.size, params.dataInputCost),
+      Math.multiplyExact(unsignedTx.outputCandidates.size, params.outputCost)
+    )
+    val maxCost = params.maxBlockCost
+    val startCost = addCost(baseCost, initialCost, maxCost, msg = unsignedTx.toString())
 
     val transactionContext = TransactionContext(boxesToSpend.map(_.box), dataBoxes, unsignedTx)
 
+    var currentCost = startCost
+
+    val (outAssets, outAssetsNum) = JavaHelpers.extractAssets(unsignedTx.outputCandidates)
+    val (inAssets, inAssetsNum) = JavaHelpers.extractAssets(boxesToSpend.map(_.box))
+
+    val tokenAccessCost = params.tokenAccessCost
+    val totalAssetsAccessCost =
+      Math.addExact(
+        Math.multiplyExact(Math.addExact(outAssetsNum, inAssetsNum), tokenAccessCost),
+        Math.multiplyExact(Math.addExact(inAssets.size, outAssets.size), tokenAccessCost))
+    currentCost = addCost(currentCost, totalAssetsAccessCost, maxCost, "")
```

Comment:
Wouldn't a message be better instead of empty string?

Reply:
Yeah, added a message.